crates/sortition/src/sortition.rs

```diff
@@ -25,84 +28,243 @@ pub struct GetHasNode {
+                registered.iter().map(|n| n.address.to_string()).collect()
+            }
         }
+    }
 
-        let registered_nodes: Vec<Address> = self
-            .nodes
-            .clone()
-            .into_iter()
-            // TODO: error handling
-            .map(|b| b.parse().unwrap())
-            .collect();
-
-        let Ok(committee) =
-            DistanceSortition::new(seed.into(), registered_nodes, size).get_committee()
-        else {
-            return Err(anyhow!("Could not get committee!"));
+    fn contains_distance(&self, seed: Seed, size: usize, address: &str) -> Result<bool> {
+        let nodes = match self {
+            SortitionBackend::Distance { nodes } => nodes,
+            _ => return Err(anyhow!("wrong backend for distance contains")),
         };
 
+        if nodes.is_empty() || size == 0 {
+            return Ok(false);
+        }
+
+        let registered_nodes: Vec<Address> = nodes
+            .iter()
+            .cloned()
+            .map(|b| b.parse::<Address>())
+            .collect::<Result<_, _>>()?;
+
+        let committee =
+            DistanceSortition::new(seed.into(), registered_nodes, size).get_committee()?;
+
         Ok(committee
             .iter()
             .any(|(_, addr)| addr.to_string() == address))
     }
 
+    fn contains_score(&self, seed: Seed, size: usize, address: &str) -> Result<bool> {
+        let registered = match self {
+            SortitionBackend::Score { registered } => registered,
+            _ => return Err(anyhow!("wrong backend for score contains")),
+        };
+
+        if registered.is_empty() || size == 0 {
+            return Ok(false);
+        }
+
+        let committee = ScoreSortition::new(size).get_committee(seed.into(), registered)?;
+        let want: Address = address.parse()?;
+        Ok(committee.iter().any(|w| w.address == want))
+    }
+}
+
+/// Per-chain module: stores the chosen algorithm and its backend storage.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SortitionModule {
+    sortition: SortitionType,
+    backend: SortitionBackend,
+}
+
+impl Default for SortitionModule {
+    fn default() -> Self {
+        Self {
+            sortition: SortitionType::Distance,
+            backend: SortitionBackend::default(),
+        }
+    }
+}
+
+impl SortitionModule {
+    /// Create a module with a specific algorithm.
+    pub fn new(sortition: SortitionType) -> Self {
+        let backend = match sortition {
+            SortitionType::Distance => SortitionBackend::Distance {
+                nodes: HashSet::new(),
+            },
+            SortitionType::Score => SortitionBackend::Score {
+                registered: Vec::new(),
+            },
+        };
+        Self { sortition, backend }
+    }
+
+    pub fn sort_type(&self) -> SortitionType {
+        self.sortition
+    }
+
+    /// Switch algorithm and migrate storage shape.
+    pub fn set_sort_type(&mut self, sort_type: SortitionType) {
```

Comment:
Seems like we need to think about this before implementing a switchover algorhythm. I think we should not support this at this point until we have a live system that actually needs updating on the fly.